File: python/demo-function-call.py
import json
from dotenv import load_dotenv
from openai import OpenAI

load_dotenv(dotenv_path="../.env")

client = OpenAI()

messages = [
    {"role": "system", "content": "あなたは親切なアシスタントです。"},
    {"role": "user", "content": "おうし座の今日の運勢を教えてください。"},
]

tools = [
    {
        "type": "function",
        "function": {
            "name": "get_horoscope",
            "description": "指定された星座の運勢を取得する",
            "parameters": {
                "type": "object",
                "properties": {
                    "sign": {
                        "type": "string",
                        "description": "星座名（例: 'おうし座'）",
                    },
                    "day": {
                        "type": "string",
                        "enum": ["today", "tomorrow"],
                        "description": "運勢を取得する日（今日または明日）",
                    },
                },
                "required": ["sign"],
                "additionalProperties": False,
            },
            "strict": True,
        },
    }
]

response = client.chat.completions.create(
    model="gpt-5-mini",
    messages=messages,
    tools=tools,
)

message = response.choices[0].message
print("finish_reason:", response.choices[0].finish_reason)

if message.tool_calls:
    for tool_call in message.tool_calls:
        func_name = tool_call.function.name
        func_args = json.loads(tool_call.function.arguments)
        print(f"tool_call: {func_name}({func_args})")

        if func_name == "get_horoscope":
            tool_result = {
                "horoscope": f"今日の{func_args.get('sign')}の運勢: 全体運は好調です！積極的に行動すると吉。"
            }
        else:
            raise RuntimeError(f"Unknown function: {func_name}")

        messages.append(message)
        messages.append(
            {
                "role": "tool",
                "tool_call_id": tool_call.id,
                "content": json.dumps(tool_result, ensure_ascii=False),
            }
        )

    response2 = client.chat.completions.create(
        model="gpt-5-mini",
        messages=messages,
    )

    print(response2.choices[0].message.content)
else:
    print(message.content)

File: nodejs/demo-function-call.mjs
import OpenAI from "openai";

const client = new OpenAI(); // OPENAI_API_KEY 環境変数を自動で読み取る

const messages = [
    { role: "system", content: "You are a helpful assistant." },
    { role: "user",   content: "Give me my horoscope for Taurus." },
];

const tools = [
    {
        type: "function",
        function: {
            name: "get_horoscope",
            description: "Get the horoscope for a given zodiac sign",
            parameters: {
                type: "object",
                properties: {
                    sign: { type: "string", description: "Zodiac sign, e.g. 'Taurus'" },
                    day:  { type: "string", enum: ["today", "tomorrow"], description: "Which day to get the horoscope for" },
                },
                required: ["sign"],
                additionalProperties: false,
            },
        },
    },
];

// ステップ1〜2: ツール定義を添えてモデルへ送信する
const response = await client.chat.completions.create({
    model: "gpt-4o-mini",
    messages,
    tools,
    tool_choice: "auto",
});

const message = response.choices[0].message;
console.log("finish_reason:", response.choices[0].finish_reason);

// ステップ3〜4: tool_callsが返ってきたらアプリ側で処理して結果を返す
if (message.tool_calls) {
    for (const toolCall of message.tool_calls) {
        const funcName = toolCall.function.name;
        const funcArgs = JSON.parse(toolCall.function.arguments);
        console.log(`tool_call: ${funcName}(${JSON.stringify(funcArgs)})`);

        let toolResult;
        if (funcName === "get_horoscope") {
            // ダミーの実行結果（実務では外部APIやDBを呼び出す）
            toolResult = { horoscope: `Today's ${funcArgs.sign} horoscope: Your fortune is looking great!` };
        } else {
            throw new Error(`Unknown function: ${funcName}`);
        }

        messages.push(message);
        messages.push({
            role: "tool",
            tool_call_id: toolCall.id,
            content: JSON.stringify(toolResult),
        });
    }

    // ステップ5: 最終応答を取得する
    const response2 = await client.chat.completions.create({
        model: "gpt-4o-mini",
        messages,
    });

    console.log(response2.choices[0].message.content);
} else {
    console.log(message.content);
}
